modules/adpod.js

```diff
@@ -0,0 +1,364 @@
+  let pcd = `${cpmFixed}_${category}_${duration}s`;
+
+  if (!bid.adserverTargeting) {
+    bid.adserverTargeting = {};
+  }
+  bid.adserverTargeting[TARGETING_KEY_PB_CAT_DUR] = pcd;
+  bid.adserverTargeting[TARGETING_KEY_CACHE_ID] = initialCacheKey;
+  bid.customCacheKey = `${pcd}_${initialCacheKey}`;
+}
+
+/**
+ * Updates the running queue for the associated auction.
+ * Does a check to ensure the auction is still running; if it's not - the previously running queue is killed.
+ * @param {*} auctionInstance running context of the auction
+ * @param {Object} bidResponse bid object being added to queue
+ * @param {Function} afterBidAdded callback function used when Prebid Cache responds
+ */
+function updateBidQueue(auctionInstance, bidResponse, afterBidAdded) {
+  let bidListIter = bidCacheRegistry.getBids(bidResponse);
+
+  if (bidListIter) {
+    let bidListArr = from(bidListIter);
+    let callDispatcher = bidCacheRegistry.getQueueDispatcher(bidResponse);
+    let killQueue = !!(auctionInstance.getAuctionStatus() !== AUCTION_IN_PROGRESS);
+    callDispatcher(auctionInstance, bidListArr, afterBidAdded, killQueue);
+  } else {
+    utils.logWarn('Attempted to cache a bid from an unknown auction. Bid:', bidResponse);
+  }
+}
+
+/**
+ * Small helper function to remove bids from internal storage; normally b/c they're about to sent to Prebid Cache for processing.
+ * @param {Array[Object]} bidResponses list of bids to remove
+ */
+function removeBidsFromStorage(bidResponses) {
+  for (let i = 0; i < bidResponses.length; i++) {
+    bidCacheRegistry.removeBid(bidResponses[i]);
+  }
+}
+
+/**
+ *
+ * @param {*} auctionInstance running context of the auction
+ * @param {Array[Object]} bidList list of bid objects that need to be sent to Prebid Cache
+ * @param {Function} afterBidAdded callback function used when Prebid Cache responds
+ */
+function firePrebidCacheCall(auctionInstance, bidList, afterBidAdded) {
+  // note to reviewers - was doing this to ensure the variable wouldn't be malformed by the additional incoming bids
+  // while the current list is being processed by PBC and the store callback
+  // ...but is this copy really needed?
+  let bidListCopy = bidList.slice();
+
+  // remove entries now so other incoming bids won't accidentally have a stale version of the list while PBC is processing the current submitted list
+  removeBidsFromStorage(bidListCopy);
+
+  store(bidListCopy, function (error, cacheIds) {
+    if (error) {
+      utils.logWarn(`Failed to save to the video cache: ${error}. Video bid(s) must be discarded.`);
+      for (let i = 0; i < bidList.length; i++) {
+        doCallbacksIfTimedout(auctionInstance, bidList[i]);
+      }
+    } else {
+      for (let i = 0; i < cacheIds.length; i++) {
+        // when uuid in response is empty string then the key already existed, so this bid wasn't cached
+        // TODO? - should we throw warning here?  should we call the afterBidAdded() anyway to decrement the internal bid counter?
```

Comment:
We should call afterBidAdded to decrement the counter because there can be a case where auction is not timedout and one of the bid is back with empty uuid. In this case we would like to close the auction and not wait for timer to expire

Reply:
To confirm on this point - should we throw any type of message that the bid wasn't cached b/c it had a duplicate key?  Or just let it go silently?

Reply:
yep logMessage is good